opdatasigverify.md

```diff
@@ -0,0 +1,122 @@
+
+The algorithm first computes the double-sha256 hash of the byte string "Bitcoin Signed Message:\n" prepended to the supplied *data* (stack top - 2).  It then computes a pubkey from this hash and the provided signature (stack top - 1, without the first byte).  If the pubkey cannot be recovered, the script fails.  It then compares the hash160 (same as OP_HASH160) of the recovered pubkey to the provided *pubkeyhash* (stack top).  If the comparison fails, the script fails.
+
+Otherwise, the top 2 items are popped off the stack (leaving *data* on the top of the stack), and the opcode succeeds.
+
+## Examples
+
+* Simplest DATASIGVERIFY:
+  * output script:  `pubkeyhash OP_DATASIGVERIFY`
+  * spend script: `data sig`
+
+* Realistic example: Verify that `data` is signed by `dpubkeyhash` (signature `daddrsig`) and equals `x`.  Also do normal p2pkh transaction verification on `pubkeyhash` (with signature `txSig` and public key `pubkey`):
+  * output script: `dpubkeyhash OP_DATASIGVERIFY x OP_EQUALVERIFY OP_DUP OP_HASH160 pubkeyhash OP_EQUALVERIFY OP_CHECKSIGVERIFY
+  * spend script: `txSig pubkey data daddrsig`
+
+## Discussion
+
+### Why is the non-verify version unnecessary?
+
+OP_DATASIGVERIFY fails the script if signature validation fails.  But one could imagine another opcode, let's call it OP_CHECKDATASIG, that would instead push true or false onto the stack depending on whether signature validation suceeded or failed.  Therefore additional opcode would relate to OP_DATASIGVERIFY in the same way OP_CHECKSIG is to OP_CHECKSIGVERIFY.
+
+But, this instruction is not necessary because a Bitcoin Cash script is not exactly like a normal program.  Its fundamental and only purpose is to encumber spendability -- your input script must meet the output script's requirements in order to spend the coins.  Therefore, any program sequence that does not narrow the group of possible spenders is equivalent to a simple choice.  In the OP_CHECKDATASIG case, any potential spender can create a spend script that fails OP_CHECKDATASIG by pushing random bytes rather than a valid signature, so the "false" result is equivalent to (but less efficient than) an if statement on a pushed constant.
+
+Formally, any script with encumberances x, y, z and solutions x', y', and z', of the form:
+
+`pubkeyhash OP_CHECKDATASIG OP_IF x OP_ELSE y OP_ENDIF z`
+
+if-branch spend: `z' x' data sig`
+
+else-branch spend: `z' y' data badsig`
+
+Can be replaced by the script:
+
+`OP_IF pubkeyhash OP_DATASIGVERIFY x OP_ELSE y OP_ENDIF z`
+
+if-branch spend:  `z' x' data sig 1`
+
+else-branch spend: `z' y' 0`
+
+
+## Reference Implementation
+
+Please refer to [this github branch](https://github.com/gandrewstone/BitcoinUnlimited/tree/op_datasigverify) for a complete implementation.  Implementation is in src/test/interpreter.cpp/h and unit tests are located at src/test/script_tests.cpp.
```

Comment:
OP_DATASIGVERIFY has been merged in BU `dev` branch (disabled state) on Apr 10th, 2018. Therefore I guess we could use that one as reference implementation. (https://github.com/BitcoinUnlimited/BitcoinUnlimited/commit/03318e9f0e2f876b1d1f4ef7eeabb37e4392e56d)